File: 27-Interpreter/src/go/and_expression.go
package main

// AndExpression is a Non-terminal Expression — logical AND of two sub-expressions.
type AndExpression struct {
	left  Expression
	right Expression
}

func NewAndExpression(left, right Expression) *AndExpression {
	return &AndExpression{left: left, right: right}
}

func (a *AndExpression) Interpret(context *Context) bool {
	return a.left.Interpret(context) && a.right.Interpret(context)
}

File: 27-Interpreter/src/rust/and_expression.rs
use crate::expression::Expression;
use crate::context::Context;

/// Non-terminal Expression — logical AND of two sub-expressions.
pub struct AndExpression {
    left: Box<dyn Expression>,
    right: Box<dyn Expression>,
}

impl AndExpression {
    pub fn new(left: Box<dyn Expression>, right: Box<dyn Expression>) -> Self {
        AndExpression { left, right }
    }
}

impl Expression for AndExpression {
    fn interpret(&self, context: &Context) -> bool {
        self.left.interpret(context) && self.right.interpret(context)
    }
}
